### ISSUE TITLE
Validate duration and cron schedule conflict in webhook

### ISSUE THREAD

**Issue by at15** [pingcap/chaos-mesh#373]
## Feature Request

**Is your feature request related to a problem? Please describe:**

From https://github.com/pingcap/chaos-mesh/pull/353#discussion_r396348890

When user give a duration that is smaller than cron interval in networkchaos.
Controller generates warning `nextRecover shouldn't be later than nextStart`

```yaml
  duration: "10s"
  scheduler:
    cron: "@every 5s
```

**Describe the feature you'd like:**

Validate duration and cron schedule in webhook instead of at runtime inside controller.
Reject invalid spec directly when user submit it.

**Describe alternatives you've considered:**

N/A

**Teachability, Documentation, Adoption, Migration Strategy:**

Should mention in doc that user should be aware of the different duration they are specifying to avoid conflict. I think even delay could cause conflict (unexpected behavior), i.e. a delay of 5s with a cron of 1s

**Comment by at15** [comment]
It seems it's not limit to network chaos, there is a `ValidateScheduler` interface for each chaos and most implementations are same, only checks at least one of duration and scheduler is presented.

https://github.com/pingcap/chaos-mesh/blob/master/api/v1alpha1/iochaos_webhook.go#L83-L95

https://github.com/pingcap/chaos-mesh/blob/master/api/v1alpha1/networkchaos_webhook.go#L106-L119

Maybe we should merge them and use a default implementation for most chaos unless they have different requirement when it comes to duration and schedule.

**Comment by cwen0** [comment]
> Maybe we should merge them and use a default implementation for most chaos unless they have different requirement when it comes to duration and schedule. 

@at15 This is a good idea, I've wanted to do this for a long time. Can you help us optimize this?

**Comment by YangKeao** [comment]
> 
> 
> > Maybe we should merge them and use a default implementation for most chaos unless they have different requirement when it comes to duration and schedule.
> 
> @at15 This is a good idea, I've wanted to do this for a long time. Can you help us optimize this?

I have tried to do this with embedded struct. However, the code generator seems cannot work with embedded structs well 😿

**Comment by Gallardot** [comment]
> I have tried to do this with embedded struct. However, it seems that the code generator cannot work with embedded structs well 😿

@YangKeao 

How about this ?

```
type SchedulerSpec struct {
	Cron string `json:"cron,omitempty"`
        Duration *string `json:"duration,omitempty"`
}
```

I think `Duration` and `Cron` are very relevant fields. Why not define them together？

**Comment by at15** [comment]
In @Gallardot  case we need to break all existing spec, or have some backward compatible logic to support both. (similar to the unimplemented part in #346 )

And @YangKeao even without embedding, we can still have a common implementation

````
func ValidateSchedule(duratoion *time.Duration, scheduler *string) field.ErrorList {
}

func (in *IoChaos) ValidateScheduler(root *field.Path) field.ErrorList { 
   return ValidateSchedule(in.Spec.Duration, in.Spec.Scheduler)
}
```

**Comment by Gallardot** [comment]
> In @Gallardot case we need to break all existing spec, or have some backward compatible logic to support both. (similar to the unimplemented part in #346 )
> 
> And @YangKeao even without embedding, we can still have a common implementation
> 
> ```
> func ValidateSchedule(duratoion *time.Duration, scheduler *string) field.ErrorList {
> }
> 
> func (in *IoChaos) ValidateScheduler(root *field.Path) field.ErrorList { 
>    return ValidateSchedule(in.Spec.Duration, in.Spec.Scheduler)
> }
> ```
> ```

I get your point. It's also a good ideas.

**Comment by at15** [comment]
I look a bit more into the two phase controller logic and it seems duration and cron are not the only two fields controlling the timing. There is also `NextStart` and `NextRecover`. They are included in spec  and marked as optional, are we expecting user to set them or they are just used internally by the two phase controller? If we allow user to set them, then we need to validate them as well. Otherwise we should reject them in ValidateCreate because they are meant to be updated by controller only.

https://github.com/pingcap/chaos-mesh/blob/093337c1f46cef80af4a8788101444efaff99cb8/controllers/twophase/types.go#L170-L171

https://github.com/pingcap/chaos-mesh/blob/093337c1f46cef80af4a8788101444efaff99cb8/controllers/twophase/types.go#L170-L171

Also I think in the validation func we can use the `InnerScheduler` because it has getter for all the time related fields, one drawback is returns `time.Time` instead of `*time.Time` but `IsZero` should be good

https://github.com/pingcap/chaos-mesh/blob/093337c1f46cef80af4a8788101444efaff99cb8/controllers/twophase/types.go#L35-L46

```
func ValidateSchduler(obj InnerScheduler) {
      nextStart := obj.GetNextStart()
      nextRecover := obj.GetNextRecover()
      if nextStart.After(nextRecover) {
           allErrs = append(allErrs, .....)
      }
      // duration & cron
      if duration > cron.(ConstantDelaySchedule) {
           allErrs
     }
}
```

**Comment by Gallardot** [comment]
> In @Gallardot case we need to break all existing spec, or have some backward compatible logic to support both. (similar to the unimplemented part in #346 )
> 
> And @YangKeao even without embedding, we can still have a common implementation
> 
> ```
> func ValidateSchedule(duratoion *time.Duration, scheduler *string) field.ErrorList {
> }
> 
> func (in *IoChaos) ValidateScheduler(root *field.Path) field.ErrorList { 
>    return ValidateSchedule(in.Spec.Duration, in.Spec.Scheduler)
> }
> ```
> ```


@at15
I refer to your suggestions, I have used it on this pr #387 . Thank you.

**Comment by at15** [comment]
I think @cwen0 you can assign this issue to @Gallardot since he is already fixing it in #387

**Comment by Gallardot** [comment]
> I think @cwen0 you can assign this issue to @Gallardot since he is already fixing it in #387

Not yet fixing this issue, It is better to fixing this issue in another PR. 😂
@cwen0 @at15

**Comment by Gallardot** [comment]
I did some experiments.I think it's OK.  I will create a PR again after #387 merged.
